src/backend/gl/src/factory.rs

```diff
@@ -382,42 +472,57 @@ impl d::Factory<R> for Factory {
         self.share.handles.borrow_mut().make_sampler(sam, info)
     }
 
-    fn map_buffer_raw(&mut self, buf: &handle::RawBuffer<R>,
-                      access: f::MapAccess) -> RawMapping {
+    fn map_buffer_raw(&mut self, buf: &handle::RawBuffer<R>, access: mapping::Access)
+                      -> Result<handle::RawMapping<R>, mapping::Error> {
         let gl = &self.share.context;
-        let raw_handle = *self.frame_handles.ref_buffer(buf);
-        unsafe { gl.BindBuffer(gl::ARRAY_BUFFER, raw_handle) };
-        let ptr = unsafe { gl.MapBuffer(gl::ARRAY_BUFFER, match access {
-            f::MapAccess::Readable => gl::READ_ONLY,
-            f::MapAccess::Writable => gl::WRITE_ONLY,
-            f::MapAccess::RW => gl::READ_WRITE
-        }) } as *mut ::std::os::raw::c_void;
-        RawMapping {
-            pointer: ptr,
-            target: gl::ARRAY_BUFFER
-        }
-    }
+        let raw_handle = *buf.resource();
+
+        try!(buf.valid_access(access));
+
+        let target = gl::ARRAY_BUFFER;
+        let (kind, ptr) = if self.share.private_caps.buffer_storage_supported {
+            let access = access_to_map_bits(access) |
+                         gl::MAP_PERSISTENT_BIT |
+                         gl::MAP_FLUSH_EXPLICIT_BIT;
```

Comment:
this bit is not necessary when there is no write access, not sure if it matters though.